Trees: use Type.FuncParamClause in Type.FunctionType

kitbellew opened the pull request:
Helps with #2869.

tgodzik commented on scalameta/trees/shared/src/main/scala/scala/meta/Trees.scala:
```diff
@@ -223,15 +223,35 @@ object Type {
   }
-  @ast class Function(params: List[Type], res: Type) extends FunctionType
+  @ast class Function(param: Type, res: Type) extends FunctionType {
+    @replacedField("4.6.0", FunctionType.paramsToParam) final override def params: List[Type] =
```
Could you maybe add some docs about it and maybe add @deprecated flag? Otherwise users are getting both `param` and `params` when doing `.` completions on the tree node for Type.Function

tgodzik commented on scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala:
```diff
@@ -629,29 +633,36 @@ class ScalametaParser(input: Input)(implicit dialect: Dialect) { parser =>
-      case _ => autoEndPos(lpPos)(tuple(body))
-    }
+  private def makeTuple[A <: Tree](lpPos: Int, body: List[A], zero: => A, tuple: List[A] => A)(
+      single: A => Either[List[A], A]
```
Couldn't this function be a Boolean?
```suggestion
      isSingleTuple: A => Boolean
```
and/or maybe add some comment what it does?

tgodzik commented on scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala:
```diff
@@ -629,29 +633,36 @@ class ScalametaParser(input: Input)(implicit dialect: Dialect) { parser =>
-      case singleArg :: Nil => maybeAnonymousLambda(singleArg) :: Nil
-      case _ => body
+  private def makeTupleType(lpPos: Int, body: List[Type], zero: => Type, wrap: Boolean): Type = {
+    makeTuple(lpPos, body, zero, Type.Tuple.apply) {
```
Everywhere aside from here we just use t, so this is the only place we need to return anything aside from Boolean (so an Either), so it seems like we are complicating the signature of the `single` function just for this one case.

kitbellew replied:
what would you suggest? either we do it like this, or we duplicate the Quasi checks instead.

kitbellew commented:
@tgodzik

- moved tuple handling logic into #2964
- changed to explicit `FuncParamClause` type rather than `Unit/Type/Tuple` depending on the situation, which also allows proper quasiquotes usage.

tgodzik commented on scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala:
```diff
@@ -226,6 +226,10 @@ class ScalametaParser(input: Input)(implicit dialect: Dialect) { parser =>
     inParensAfterOpen(body)
   }
+  @inline final def inParensOr[T](body: => T)(alt: => T): T = {
+    accept[LeftParen]
+    inParensAfterOpenOr(body)(alt)
```
```suggestion
    if (acceptOpt[LeftParen]) inParensAfterOpenOr(body)(alt)
    else alt
```
?
Otherwise, the method seems misleading since it doesn't accept no parens.

kitbellew replied:
for both comments: the method requires _parens_ but it doesn't require anything _in_ them (that is, allows the closing paren right away).

tgodzik replied:
Maybe then `inParens` instead `inParensOr` : 
```
   @inline final def inParens[T](body: => T)(whenEmpty: => T): T = {
```

kitbellew replied:
there is already a method called that way, for cases when there's no "or", meaning that an immediate closing paren is not expected... let me review.

kitbellew replied:
so, here's what we have:

- both parameters (_body_ and _alt_) are of the same type, so we can't reorder them to disambiguate from regular `inParens`
- there are several methods named similarly (two sets right below) and behaving similarly.

kitbellew replied:
thanks, will rename the `alt` parameter.

kitbellew replied:
added a separate PR for that, #2967.

kitbellew replied:
@tgodzik that PR has been merged. please let me know if there are any other adjustments to make.

tgodzik commented on scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala:
```diff
@@ -716,14 +720,7 @@ class ScalametaParser(input: Input)(implicit dialect: Dialect) { parser =>
-      }
-      else Nil
-      accept[RightParen]
+      val ts = inParensOr(commaSeparated(paramOrType()))(Nil)
```
```suggestion
      val ts = inParensOr(commaSeparated(paramOrType()))(syntaxError(...))
```
? since we expect them and we can give a meaningful error message here? Parens are expected and if there are not there we should show error.

tgodzik reviewed (changes requested):
Just two small comments, thanks for the other PRs!

tgodzik reviewed (approved):
LGTM
